ledger-service/http-json/search-query-language.rst

```diff
@@ -0,0 +1,99 @@
+<../lf-value-json/specification.rst>`_, and compared for equality.
+
+Simple equality
+---------------
+
+Match documents having at least all the (potentially nested) keys
+expressed in the query. The result document may contain additional
+properties.
+
+Example: ``{ person: { name: "Bob" }, city: "London" }``
+
+- Match: ``{ person: { name: "Bob", dob: "1956-06-21" }, city: "London",
+  createdAt: "2019-04-30T12:34:12Z" }``
+- No match: ``{ person: { name: "Bob" }, city: "Zurich" }``
+- Typecheck failure: ``{ person: { name: ["Bob", "Sue"] }, city:
+  "London" }``
+
+Example: ``{ favorites: ["vanilla", "chocolate"] }``
+
+- Match: ``{ favorites: ["vanilla", "chocolate"] }``
+- No match: ``{ favorites: ["chocolate", "vanilla"] }``
+- No match: ``{ favorites: ["vanilla", "strawberry"] }``
+- No match: ``{ favorites: ["vanilla", "chocolate", "strawberry"] }``
+
+A JSON object, when considered with a record type, is always interpreted
+as a field equality query. Its type context is thus mutually exclusive
+with `the forthcoming comparison queries
+<https://github.com/digital-asset/daml/issues/2780>`_.
+
+Appendix: Type-aware queries
+----------------------------
+
+**This section is non-normative.**
+
+This is not a *JSON* query language, it is a *DAML-LF* query
+language. So, while we could theoretically treat queries (where not
+otherwise interpreted by the "may contain additional properties" rule
+above) without concern for what LF type (i.e. template) we're
+considering, we *will not* do so.
+
+Consider the subquery ``{"foo": "bar"}``. This query conforms to types,
+among an unbounded number of others::
+
+  record A ↦ { foo : Text }
+  record B ↦ { foo : Optional Text }
+  variant C ↦ foo : Party | bar : Unit
+
+  // NB: LF does not require any particular case for VariantCon or Field;
+  // these are perfectly legal types in DAML-LF packages
+
+In the cases of ``A`` and ``B``, ``"foo"`` is part of the query
+language, and only ``"bar"`` is treated as an LF value; in the case of
+``C``, the whole query is treated as an LF value. The wide variety of
+ambiguous interpretations about what elements are interpreted, and what
+elements treated as literal, and *how* those elements are interpreted or
+compared, would preclude many techniques for efficient query compilation
+and LF value representation that we might otherwise consider.
+
+Additionally, it would be extremely easy to overlook unintended meanings
+of queries when writing them, and impossible in many cases to suppress
+those unintended meanings within the query language. For example, there
+is no way that the above query could be written to match ``A`` but never
+``C``.
+
+For these reasons, as with LF value input via JSON, queries written in
+JSON are also always interpreted with respect to some specified LF types
+(e.g. template IDs). As #2777 implies, for example::
+
+  {"%templates": [{"moduleName": "Foo", "entityName": "A"},
```

Comment:
Just thinking about this... Maybe `%templates` should NOT be an `Array`? In this case if typecheck fails, it will be clear what `Template` it failed for. I bet we can improve the error reporting and maybe it already provides all required information, but it just looks strange to me that we may return totally unrelated types in one result set.

Reply:
If it helps, you can think of the result type as a list of values of variant type, where the variant data constructor is a template ID rather than a string or int. If it does not help, well, we can reconsider elsewhere.